### PR TITLE
Update UI to match Mozilla AWS CLI

### LINKED ISSUE #5: If you change roles it prompts you to logout first
Maybe have the user logout in the background each time before doing a login

### LINKED ISSUE #7: Update role picker UI to match federated-aws-cli, sorted by AWS account
This should likely be accomplished by changing to a frontend / backend model that uses the same js as the Mozilla AWS CLI.

### PR DESCRIPTION
**Pull request by gene1wood**
This changes the UI and flow of federated-aws-rp from a syncronous
backend web flow to an asyncronous frontend + backend web flow.

This allows re-using the UI from the Mozilla AWS CLI so that they
match.

A side effect of this will be to fix #5 by logging the user out before
logging them in.

To do this I've made the following changes
* Enabled sending binary data to users to support sending font files
* Added support for POSTing to the RP
* Added API endpoints to mirror the API endpoints in mozilla-aws-cli
* Add support for using account aliases in the "account" query argument
  utils.py:convert_account_alias
* Add sorting and grouping of roles by account for presentation in the
  role picker
* Added support for reading files from the package
* Included static assets (js, html, fonts) copied from mozilla-aws-cli
* Added a fake /api/heartbeat endpoint since we don't need one like maws
  does

This is a bit rewrite of the project and affects a lot of surface area.

Fixes #7

### CODE REVIEW

File: federated-aws-rp.yaml [hunk 1]
```diff
@@ -286,12 +309,17 @@ Resources:
...
           AllowedMethods:
-            - GET
-            - HEAD
+          - GET
+          - HEAD
+          - POST
+          - DELETE
+          - OPTIONS
+          - PUT
+          - PATCH
```
Comment (april) [inline federated-aws-rp.yaml:322]: Is there a reason why we need to allow these methods?

File: federated-aws-rp.yaml [hunk 2]
```diff
@@ -312,33 +340,33 @@ Resources:
...
-              OriginProtocolPolicy: https-only
-              OriginSSLProtocols:
-                - TLSv1.2
-            DomainName: !GetAtt AwsFederatedRpApiGatewayDomainName.RegionalDomainName
-            Id: AwsFederatedRpCloudFrontOriginId
-            # OriginPath: !Join [ '', [ '/', !Ref 'AwsFederatedRpApiStage' ] ]
+        - CustomOriginConfig:
+            OriginProtocolPolicy: https-only
+            OriginSSLProtocols:
+              - TLSv1.2
```
Comment (april) [inline federated-aws-rp.yaml:346]: Can we specify TLSv1.3 here?

File: federated-aws-rp.yaml [hunk 3]
```diff
@@ -312,33 +340,33 @@ Resources:
...
+          DomainName: !GetAtt AwsFederatedRpApiGatewayDomainName.RegionalDomainName
+          Id: AwsFederatedRpCloudFrontOriginId
+          # OriginPath: !Join [ '', [ '/', !Ref 'AwsFederatedRpApiStage' ] ]
```
Comment (april) [inline federated-aws-rp.yaml:349]: Can I ask what this is for?

File: functions/federated_aws_rp/app.py [hunk 4]
```diff
@@ -16,175 +17,144 @@
...
-    exchange_token_for_roles,
-    redirect_to_web_console,
-    trigger_group_role_map_rebuild
+    get_aws_federation_url,
+    trigger_group_role_map_rebuild,
+    login,
+    get_roles,
+    read_resource,
+    return_api_gateway_json,
+    convert_account_alias
```
Comment (april) [inline functions/federated_aws_rp/app.py]: Alphabetize these?

File: functions/federated_aws_rp/app.py [hunk 5]
```diff
@@ -16,175 +17,144 @@
...
-            raise AccessDenied('"state" query parameter is missing"')
-        elif 'code' is None:
-            raise AccessDenied('"code" query parameter is missing')
-        store = get_store(cookie_header)
-        if store.get('state') != state:
-            raise AccessDenied('Invalid "state" query parameter passed')
-        elif 'code_verifier' not in store:
-            raise AccessDenied(
-                '"code_verifier" not found in cookie')
+        logger.debug('/redirect_callback called with body {}'.format(body))
```
Comment (april) [inline functions/federated_aws_rp/app.py]: Maybe have this use `json.dumps` for legibility?

File: functions/federated_aws_rp/app.py [hunk 6]
```diff
@@ -16,175 +17,144 @@
...
-            'body': CONFIG.role_picker_template.render(roles=roles["roles"])
-        }
-    return response
+            store['client_workflow_value'] = 'Access denied'
+        else:
+            store['client_workflow_value'] = 'Group Role Map rebuild initiated'
+    else:
+        logger.error('Invalid action argument {}'.format(store.get('action')))
+        store['client_workflow_state'] = 'error'
+        store['client_workflow_value'] = 'Invalid action argument'
```
Comment (april) [inline functions/federated_aws_rp/app.py:134]: I see code like this a lot. Should this be a function? Like `set_store_state(state, value)`?

File: functions/federated_aws_rp/app.py [hunk 7]
```diff
@@ -16,175 +17,144 @@
...
+    store['last_state_check'] = time.time()
+    return return_api_gateway_json(
+        store,
+        {
+            'state': store.get('client_workflow_state'),
+            'value': store.get('client_workflow_value')
         },
-        'body': body}
+        False
+    )
```
Comment (april) [inline functions/federated_aws_rp/app.py:156]: This is really strange. What is `False`? Should we return `state` as a standalone thing *and* the values inside state as an dictionary?

File: functions/federated_aws_rp/app.py [hunk 8]
```diff
@@ -244,16 +224,20 @@ def redirect_to_idp(
...
+            'Content-Type': 'text/html',
             'Cache-Control': 'max-age=0',
-            'Set-Cookie': '{}={}; Secure; HttpOnly; Max-Age=3600'.format(
-                CONFIG.cookie_name,
-                encode_cookie_value(store)),
-            'Content-Type': 'text/html'
+            'Set-Cookie': (
+                '{}={}; Secure; HttpOnly; Path=/; Max-Age=3600'.format(
+                    CONFIG.cookie_name,
+                    encode_cookie_value(store)))
```
Comment (april) [inline functions/federated_aws_rp/app.py:239]: Probably should set these as `SameSite=Strict` or `SameSite=Lax` as well.

File: functions/federated_aws_rp/app.py [hunk 9]
```diff
@@ -294,24 +290,84 @@ def lambda_handler(event: dict, context: dict) -> dict:
...
+        else:
+            body = read_resource(path)
+            if body:
+                # static resource
+                content_type_map = {
+                    '.js': 'application/javascript',
+                    '.html': 'text/html',
+                    '.css': 'text/css',
+                    '.woff': 'application/octet-stream',
+                    '.woff2': 'application/octet-stream'
```
Comment (april) [inline functions/federated_aws_rp/app.py:346]: We should probably use `mimetypes`, I think.

File: functions/federated_aws_rp/app.py [hunk 10]
```diff
@@ -294,24 +290,84 @@ def lambda_handler(event: dict, context: dict) -> dict:
...
+                for extension in content_type_map:
+                    if path.endswith(extension):
+                        content_type = content_type_map[extension]
+                is_base64_encoded = type(body) == bytes
+                return {
+                    'headers': {'Content-Type': content_type},
+                    'statusCode': 200,
+                    'isBase64Encoded': is_base64_encoded,
+                    'body': (base64.b64encode(body) if is_base64_encoded
+                             else body)}
```
Comment (april) [inline functions/federated_aws_rp/app.py:358]: This seems really really complicated, I'd love to walk through this with you.

**Comment by april** [comment]
BTW, I am reviewing this but it's huge and it's taking a while.

**Review by april — COMMENTED** [review]
I have to take care of something else, so this is a partial review.

File: federated-aws-rp.yaml (hunk 1 above)
Comment (gene1wood) [reply federated-aws-rp.yaml:322]: Yes, [this field can be in one of 3 configurations](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-cloudfront-distribution-defaultcachebehavior.html#cfn-cloudfront-distribution-defaultcachebehavior-allowedmethods) and to get POST, we have to do all of these.

File: federated-aws-rp.yaml (hunk 2 above)
Comment (gene1wood) [reply federated-aws-rp.yaml:346]: Unfortunately no : https://docs.aws.amazon.com/cloudfront/latest/APIReference/API_OriginSslProtocols.html

File: federated-aws-rp.yaml (hunk 3 above)
Comment (gene1wood) [reply federated-aws-rp.yaml:349]: This is a commented out value from when federated-aws-rp didn't sit behind CloudFront. This option isn't usable any more but would be required if we didn't have CloudFront. The comment could be removed.

File: functions/federated_aws_rp/app.py (hunk 4 above)
Comment (gene1wood) [reply functions/federated_aws_rp/app.py]: Good idea, will do.

File: functions/federated_aws_rp/app.py (hunk 5 above)
Comment (gene1wood) [reply functions/federated_aws_rp/app.py]: Sure thing

File: functions/federated_aws_rp/app.py (hunk 6 above)
Comment (gene1wood) [reply functions/federated_aws_rp/app.py:134]: I would but as the functions in which these stanzas (where we set these two store variables) occur are not within a class, I don't know of a graceful way to set a variable within that functions local context by calling some other function.

I think it's ok to leave them as they are. I think the only reason it looks repeated is that the key name is long (`client_workflow_state`). Each case is setting something intentional and isn't repeated. And the reason for the verbose key name is to differentiate it from the many uses of the term `state` throughout the project.

File: functions/federated_aws_rp/app.py (hunk 7 above)
Comment (gene1wood) [reply functions/federated_aws_rp/app.py:156]: > What is False?

This is a [boolean setting telling the `return_api_gateway_json` whether or not to set the cookie value in the response](https://github.com/mozilla-iam/federated-aws-rp/pull/14/files#diff-e7248d7f4fbd48335e927de073adda41R536). We need this as sometimes when returning a  web response we want to set cookies and sometimes we don't.

> Should we return state as a standalone thing and the values inside state as an dictionary?

The `return_api_gateway_json` function takes three arguments, the cookies store, the page body, and whether or not to set the cookie store.

We can communicate the same thing by removing the `set_cookie` boolean and either passing a `store` value or passing `None`.

I've make this change in 870c2de.

File: functions/federated_aws_rp/app.py (hunk 10 above)
Comment (gene1wood) [reply functions/federated_aws_rp/app.py:358]: I've added a comment in 7cdda92 and have sent an invite for a time to chat

File: functions/federated_aws_rp/app.py (hunk 9 above)
Comment (gene1wood) [reply functions/federated_aws_rp/app.py:346]: Will do

File: functions/federated_aws_rp/app.py (hunk 8 above)
Comment (gene1wood) [reply functions/federated_aws_rp/app.py:239]: Will do

**Comment by april** [comment]
Overall things look good to me!

**Pull request merged**